tactics/btermdn.ml

```diff
@@ -76,32 +77,132 @@ let evaluable_named id env ts =
   match Structures.PrimitiveProjections.find_opt c with
-  | None -> (GRLabel (ConstRef c), stack)
+  | None -> (None, args)
   | Some p ->
-    let n_args_needed = Structures.Structure.projection_nparams c + 1 in (* +1 for the record value itself *)
-    let n_args_given = List.length stack in
+    let n_args_needed = Names.Projection.Repr.npars p + 1 in (* +1 for the record value itself *)
+    let n_args_given = List.length args in
     let n_args_missing = max (n_args_needed - n_args_given) 0 in
     let n_args_drop = min (n_args_needed - 1) n_args_given in (* we do not drop the record value from the stack *)
-    (ProjLabel (p, n_args_missing), List.skipn n_args_drop stack)
+    (Some (p, n_args_missing), List.skipn n_args_drop args)
+
+let label_of_opaque_constant c stack =
+  match decomp_constant c stack with
+  | (None, stack) -> (GRLabel (ConstRef c), stack)
+  | (Some (p, i), stack) -> (ProjLabel (p,i), stack)
+
+type constr_res = (term_label * partial_constr list) lookup_res
+and partial_constr =
+  | Constr of EConstr.t
+  | PartialConstr of constr_res
+
+let rec pr_constr_res pr_constr (cr : constr_res) =
+  let open Pp in
+  let pr_partial_constr (pc : partial_constr) =
+    match pc with
+    | Constr c ->
+      str "Constr(" ++ hv 2 (pr_constr c) ++ str ")"
+    | PartialConstr pc ->
+      str "PartialConstr([" ++ hv 2 (pr_constr_res pr_constr pc) ++ str "])"
+  in
+  let aux (lbl, pcs) =
+    pr_term_label lbl ++
+    str "," ++
+    prlist_with_sep
+      (fun () -> str ",")
+      pr_partial_constr
+      pcs
+  in
+  pr_lookup_res aux cr
+
+let decomp_lambda_constr env sigma ts decomp : EConstr.t -> EConstr.t -> constr_res =
+  let res ds p =
+    match ds with
+    | [] -> assert false
+    | ty :: ds ->
+      let acc = Label (LamLabel, [Constr ty; p]) in
+      let fn acc ty = (Label (LamLabel, [Constr ty; PartialConstr acc])) in
+      let res = List.fold_left fn acc ds in
+      res
+  in
+  let rec decomp_app c stack =
+    let k = EConstr.kind sigma c in
+    match k with
+    | App (f, args) -> decomp_app f (Array.to_list args @ stack)
+    | Const (c, _) -> (k, decomp_constant c stack)
+    | Proj (p, b, c) -> (k, (Some (Projection.repr p, 0), c :: stack))
+    | _ -> (k, (None, stack))
+  in
+  let rec go numds ds p  : constr_res =
+    match decomp_app p [] with
+    | (Lambda (_, ty, c), (None, [])) ->
+      let ds = ty :: ds in
+      go (numds + 1) ds c
+    | (_, (None, [])) ->        (* this is neither a lambda nor an application *)
+      begin
+        match decomp [] p with
+        | _ when ds = [] -> assert false
+        | Label _ ->
+          (* there are left-over lambdas and the body is discriminating *)
+          res ds (Constr p)
+        | Nothing | Everything -> Everything
+      end
+    | (_ as f, (None, args)) -> (* this is an application whose head [f] is not a projection *)
+      debug Pp.(fun () -> str "decomp_lambda_constr: not a projection");
```

Comment:
Are you sure we want this debug function in here? The output might not be very readable in practice, it seems a bit weird to put it in the decomp function. If you want to have access to debug it might be cleaner to define it as a function on the side and print the result after decomposing the term.

Reply:
I actually have a debug printer for the result. So I just removed these intermediate debugging outputs.